### PR TITLE
Hash symbol table

### PR DESCRIPTION
**Pull request by wingo**
This patch stores symbols in a simple hash table instead of a list, speeding up symbol->string.

The structure is complicated enough though that it makes sense for the garbage collector to add the symbol table to the root set.  Given that we have no static data and the heap image doesn't leak from one bootstrap to the next, this seems to be acceptable.

The improvement on bootstrap times is large:

```
Compiling stage1 compiler...
  Time: 73.408 s

Compiling stage2 compiler...
  Time: 71.36 s

Compiling stage3 compiler...
  Time: 5.733 s

Percent change (1 -> 3): -92%
Percent change (2 -> 3): -92%
```

### CODE REVIEW

**Comment by wingo** [comment]
Related issues: #32 #39

**Comment by wingo** [comment]
I should mention also that this change makes it so that gensyms are no longer interned in the symbol table.

File: schism/compiler.ss [hunk 1]
```diff
@@ -40,6 +40,7 @@
 
   (define (allocation-pointer) 0)
   (define (word-size) 4)
+  (define (symbol-table-width) 32)
```
Comment (eholk) [inline schism/compiler.ss]: nit: We should probably have a comment here saying the symbol table size must be a power of 2.

**Comment by eholk** [comment]
> I should mention also that this change makes it so that gensyms are no longer interned in the symbol table.

Honestly, I'm not sure why I was doing it that way in the first place. We weren't saving the string part of gensyms anyway, so it's not like we were getting any debugging benefits.

**Pull request merged**